data-structures.md

```diff
@@ -107,24 +109,45 @@ type Actor struct {
 	Head    Cid
     
     // Nonce is a counter of the number of messages this actor has sent
-	Nonce   Integer
+	Nonce   Uint64
     
     // Balance is this actors current balance of filecoin
-	Balance AttoFIL
+	Balance BigInteger
 }
 ```
 
+## State Tree
 
+The state trie keeps track of all state in Filecoin. It is effectively a map of addresses to `actors` in the system. It is implemented using a HAMT.
 
+## HAMT
 
-### Serialization
+TODO: link to spec for our CHAMP HAMT
 
-Actors are currently serialized simply by CBOR marshaling them, using lower-camel-cased field names.
 
-## State Tree
 
-The state trie keeps track of all state in Filecoin. It is effectively a map of addresses to `actors` in the system. It is implemented using a HAMT.
+# Filecoin Compact Serialization
 
-## HAMT
+Datastructures in Filecoin are encoded as compactly as is reasonable. At a high level, each object is converted into an ordered array of its fields (ordered by their appearance in the struct declaration), then CBOR marshaled, and prepended with an object type tag.
+
+| object | tag  |
```

Comment:
how does this encoding scheme look for multiple versions in the code? do you get a types.BlockV1 or types.BlockV2 depending on what you decoded and then have ifs? do you get a types.Block that has fields that might or might not have what you expect and you if on the version which is surfaced maybe in an unserialized field?

Reply:
practically speaking, a change in block format will happen at a fixed height, and all blocks created after that height will be the new format. disambiguating should be straightforward.

I'm not 100% sure how implementations will handle this, but i would expect a single block type with any new fields added, and a version field the code can check (or just a height check, since it will need to be aware of that anyways)